presto-main/src/main/java/io/prestosql/sql/planner/QueryPlanner.java

```diff
@@ -470,6 +474,65 @@ public DeleteNode plan(Delete node)
         return new DeleteNode(idAllocator.getNextId(), builder.getRoot(), new DeleteTarget(handle, metadata.getTableMetadata(session, handle).getTable()), rowId, outputs);
     }
 
+    public UpdateNode plan(Update node)
+    {
+        Table table = node.getTable();
+        TableHandle handle = analysis.getTableHandle(table);
+
+        Map<String, ColumnHandle> dataColumnHandles = metadata.getColumnHandles(session, handle);
+        List<String> targetColumnNames = node.getAssignments().stream()
+                .map(assignment -> assignment.getName().getValue())
+                .collect(toImmutableList());
+
+        // Create a list of column SET expressions, in table column order
+        ImmutableList.Builder<Expression> orderedColumnValuesBuilder = ImmutableList.builder();
+        for (String columnName : dataColumnHandles.keySet()) {
+            int index = targetColumnNames.indexOf(columnName);
+            if (index >= 0) {
+                orderedColumnValuesBuilder.add(node.getAssignments().get(index).getValue());
+            }
+        }
+        List<Expression> orderedColumnValues = orderedColumnValuesBuilder.build();
+
+        // create table scan
+        RelationPlan relationPlan = new RelationPlanner(analysis, symbolAllocator, idAllocator, lambdaDeclarationToSymbolMap, metadata, outerContext, session, recursiveSubqueries)
+                .process(table, null);
+
+        PlanBuilder builder = newPlanBuilder(relationPlan, analysis, lambdaDeclarationToSymbolMap);
+
+        if (node.getWhere().isPresent()) {
+            builder = filter(builder, node.getWhere().get(), node);
+        }
+
+        builder = builder.appendProjections(orderedColumnValues, symbolAllocator, idAllocator);
+
+        PlanAndMappings planAndMappings = coerce(builder, orderedColumnValues, analysis, idAllocator, symbolAllocator, typeCoercion);
+        builder = planAndMappings.getSubPlan();
+
+        ImmutableList.Builder<Symbol> updatedColumnValuesBuilder = ImmutableList.builder();
+        orderedColumnValues.forEach(columnValue -> updatedColumnValuesBuilder.add(planAndMappings.get(columnValue)));
+        Symbol rowId = builder.translate(analysis.getRowIdField(table));
+        updatedColumnValuesBuilder.add(rowId);
+        List<Symbol> updatedColumnValues = updatedColumnValuesBuilder.build();
+
+        Map<String, ColumnHandle> updatedColumns = dataColumnHandles.entrySet().stream()
```

Comment:
In https://github.com/prestosql/presto/pull/5861#discussion_r530662787, you mentioned that the list of columns in UpdateTarget needs to have the same order as `updatedColumnValues`. However, now this is a `Map`, so the order is not guaranteed.

Reply:
I'm glad you pushed on this @martint!  I rewrote this code to use `Lists` exclusively, avoiding the `Map` order hazard.  I also did some renaming to make the usage more obvious.